app/Commands/NewCommand.php

```diff
@@ -0,0 +1,474 @@
+<?php
+
+namespace App\Commands;
+
+use Illuminate\Filesystem;
+use Symfony\Component\Process\Process;
+use LaravelZero\Framework\Commands\Command;
+use Symfony\Component\Process\ExecutableFinder;
+use Symfony\Component\Process\Exception\ProcessFailedException;
+
+class NewCommand extends Command
+{
+    /**
+     * The name and signature of the console command.
+     *
+     * @var string
+     */
+    protected $signature = 'new
+                        {name : Name of the Laravel project}
+                        {--a|auth : Run make:auth}
+                        {--b|browser : Browser you want to open the project in}
+                        {--c|createdb= : Create a database; pass in sqlite or mysql}
+                        {--d|dev  : Choose the dev branch instead of master}
+                        {--e|editor= : Text editor to open the project in}
+                        {--l|link : Create a Valet link to the project directory}
+                        {--m|message= : Set the first commit message}
+                        {--y|node : Set to execute yarn or npm install}
+                        {--p|path= : Base path for the installation, otherwise CWD is used}';
+
+    protected $projectname = '';
+
+    protected $projectpath = '';
+
+    protected $projecturl = '';
+
+    protected $cwd = '';
+
+    protected $basepath = '';
+
+    protected $commitmessage = 'Initial commit.';
+
+    protected $editors_terminal = ['vim', 'vi', 'nano', 'pico', 'ed', 'emacs', 'nvim'];
+
+    protected $editors_gui = ['pstorm', 'subl', 'atom', 'textmate', 'geany'];
+
+    protected $editor = '';
+
+    protected $tools = [];
+
+    protected $dbtypes = ['sqlite', 'mysql'];
+
+    protected $os = null;
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Creates a new Laravel application';
+
+    /**
+     * Create a new command instance.
+     */
+    public function __construct()
+    {
+        parent::__construct();
+
+        $this->cwd = getcwd();
+        $this->basepath = $this->cwd;
+
+        $this->os = new \Tivie\OS\Detector();
+    }
+
+    /**
+     * Execute the console command.
+     *
+     * @return mixed
+     */
+    public function handle(): void
+    {
+        $this->projectname = $this->argument('name');
+        $this->projecturl = 'http://'.$this->projectname.'.dev';
+
+        $this->getAvailableTools();
+        if (! $this->tools['laravel']) {
+            $this->error('Unable to find laravel installer so I must exit. One day I will use composer here instead of exiting.');
+            exit;
+        }
+
+        $this->setBasePath();
+
+        $this->projectpath = $this->basepath.DIRECTORY_SEPARATOR.$this->projectname;
```

Comment:
I'm not sure the best way to do that considering the value depends on the optional `--path`, perhaps you have suggestions?

Reply:
Sure! It's basically just more tokenization and just-in-time processing. Basically there are two main ways to handle this sort of thing; we can either run a bunch of setters (if we do that, I like to do it in a constructor so we're not constantly mutating the state of the object) or we can access the data as if it's already been set and let the accessor (this method) handle setting it if it's not set yet. Six of one half dozen of another as to which is the better solution in any given context.

```php
    public function basepath()
    {
        if (! $this->basepath) {
             // This block could also just be accessing `$this->setBasePath()` instead of inlining it
             $this->basepath = $this->cwd;
         
            if ($this->option('path')) {
                $path = $this->option('path');
                if (is_dir($path)) {
                    $this->basepath = $path;
                } else {
                    $this->warn("Your defined '--path $path' is not a directory, so I am skipping it and using '{$this->basepath}' instead.");
                }
            }       
        }

        return $this->basepath;
    }
```

The other option is to handle it up front; if so, I'd like to try to mush all the setters together: 

```php
public function handle()
{
    $this->setup();

    // do the rest of the command that assumes it has access to things like $this->basepath
}

protected function setup()
{
    // set this->basepath, etc.
}
```

Reply:
Thanks for this! I added setup() (commit 3ce636b) and moved a few variable definitions there.